arwen/host/execution.go

```diff
@@ -619,21 +619,27 @@ func (host *vmHost) callSCMethodIndirect() error {
 	return err
 }
 
+// RevertESDTTransfer calls the ESDT/ESDTNFT transfer with reverted arguments
 func (host *vmHost) RevertESDTTransfer(input *vmcommon.ContractCallInput) {
```

Comment:
Would it be possible to change the `ESDTTransfer` built-in function to add a special `OutputTransfer` to the `VMOutput` instead of writing to the state? This would make `RevertESDTTransfer()` unnecessary, as the `OutputTransfer` would be cancelled with the entire `VMOutput`.

Reply:
Would need to implement treating the output transfers even intra shard in case of sc processor. It makes sense to take it out in one hand, but it complicates in the other. And you need to have a lot of extra protection on different levels which function is allowed to do what. In case of builtinfunctions, only they are allowed to execute and operate with the data under the protected keys.